src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -718,6 +770,105 @@ private static bool IsW3CId(string id)
+        public void SetCustomProperty(string propertyName, object? propertyValue)
+        {
+            // We don't check null name here as the dictionary is performing this check anyway.
+
+            if (_customProperties == null)
+            {
+                Interlocked.CompareExchange(ref _customProperties, new ConcurrentDictionary<string, object>(), null);
+            }
+
+            _customProperties[propertyName] = propertyValue!;
+        }
+
+        /// <summary>
+        /// GetCustomProperty retrieve previously attached object mapped to the property name.
+        /// </summary>
+        /// <param name="propertyName"> The name to get the associated object with.</param>
+        /// <returns>The object mapped to the property name. Or null if there is no mapping previously done with this property name.</returns>
+        public object? GetCustomProperty(string propertyName)
+        {
+            // We don't check null name here as the dictionary is performing this check anyway.
+
+            if (_customProperties == null)
+            {
+                return null;
+            }
+
+            return _customProperties.TryGetValue(propertyName, out object? o) ? o! : null;
+        }
+
+        internal static Activity CreateAndStart(ActivitySource source, ActivityContext context)
+        {
+            Debug.Assert(source != null);
+
+            Activity activity = new Activity(source.Name);
+
+            activity._source = source;
+
+            Activity? parent = Current;
+            if (parent != null)
+            {
+                // The parent change should not form a loop. We are actually guaranteed this because
+                // 1. Un-started activities can't be 'Current' (thus can't be 'parent'), we throw if you try.
+                // 2. All started activities have a finite parent change (by inductive reasoning).
+                activity.Parent = parent;
+            }
+
+            activity.IdFormat =
+                ForceDefaultIdFormat ? DefaultIdFormat :
+                activity.Parent != null ? activity.Parent.IdFormat :
+                activity._parentSpanId != null ? ActivityIdFormat.W3C :
+                activity._parentId == null ? DefaultIdFormat :
+                IsW3CId(activity._parentId) ? ActivityIdFormat.W3C :
+                ActivityIdFormat.Hierarchical;
+
+            if (context.IsValid)
+            {
+                activity._traceId = context.TraceId.ToString();
+                activity._spanId = context.SpanId.ToString();
+                activity.ActivityTraceFlags = context.TraceFlags;
+                activity._traceState = context.TraceState;
+            }
+            else
+            {
+                if (activity.IdFormat == ActivityIdFormat.W3C)
+                    activity.GenerateW3CId();
+                else
+                    activity._id = activity.GenerateHierarchicalId();
+            }
+
+            activity.StartTimeUtc = GetUtcNow();
```

Comment:
OTel allows to optionally specify start time.

Reply:
I'll add the overload in ActivitySource.CreateActivity that takes start time. would that address this point? or you want the activity to expose a property to set it (which I don't think we need that)?

Reply:
I have added the overload. please let me know if it looks ok to you.